src/box/xrow.h

```diff
@@ -355,9 +354,14 @@ struct synchro_request {
 	int64_t lsn;
 	/**
 	 * The new term the instance issuing this request is in. Only used for
-	 * PROMOTE request.
+	 * PROMOTE and DEMOTE requests.
 	 */
 	uint64_t term;
+	/**
+	 * Confirmed lsns of all the previous limbo owners. Only used for
+	 * PROMOTE and DEMOTE requests.
+	 */
+	struct vclock *confirmed_vclock;
```

Comment:
Having the single `synchro_request` for all request types was fine, but since now it starts looking as ugly as the old huge subscribe request was looking with all its fields being optional and it being used for all requests at once. I would suggest to split the request into separate `synchro_ownership_request` (for `PROMOTE` and `DEMOTE`) and `synchro_txn_request` (for `CONFIRM`/`ROLLBACK`). The latter wouldn't have `confirmed_vclock` field. Some or all the other fields could be duplicated, yes, but it would look simpler.

Reply:
I've thought about this for some time and I think it would turn out quite ugly.

First of all, I'd like to keep the existing synchro request processing code intact. This means, the ownership request should be inherited from the base synchro request. So that I can pass either of them by pointer and let the processing code (`txn_limbo_req_prepare`/`txn_limbo_process`/`txn_limbo_req_commit`) decide how to interpret the input. This particular part would work ok.

Now, if I inherit the synchro ownership requrest from the base request, I want to have a single encoder/decoder for their common part. To avoid a case when we have to change two separate decoders when the base request is changed somehow.

I tried to implement these parts, and they look satisfactory. Not perfect, but ok. The main problem is that I either end up with going over the same request body twice (can be seen in patch below) or end up with two almost identical decoders/encoders (what I tried to avoid). Anyway, I attached the code below and IMO it already is questionable. And there are problems besides encoding/decoding which we haven't talked about yet.

There's one more possible approach: create a proxy struct holding all the possible synchro request field pointers, like you did with `struct replication_request`. But then we end up with the same decoder which accepts every possible field and every field is optional. I think this is what you wanted to avoid. So, it doesn't solve the problem, but adds complications of splitting request into two, just like the other approaches.

<details><summary>Here's what I've tried</summary>

```diff
diff --git a/src/box/xrow.c b/src/box/xrow.c
index 78ee40450..49fa04fc7 100644
--- a/src/box/xrow.c
+++ b/src/box/xrow.c
@@ -1237,37 +1237,94 @@ xrow_encode_id(struct xrow_header *row)
 	row->body[0].iov_len = size;
 }
 
+uint32_t
+xrow_encode_synchro_body(char **pos, const struct synchro_request *req)
+{
+	/* Skip 1 byte for the yet unknown map size. */
+	++*pos;
+
+	uint32_t map_size = 0;
+	*pos = mp_encode_uint(*pos, IPROTO_REPLICA_ID);
+	*pos = mp_encode_uint(*pos, req->replica_id);
+	map_size++;
+	*pos = mp_encode_uint(*pos, IPROTO_LSN);
+	*pos = mp_encode_uint(*pos, req->lsn);
+	map_size++;
+	return map_size;
+}
+
 void
-xrow_encode_synchro(struct xrow_header *row, char *body,
-		    const struct synchro_request *req)
+xrow_encode_synchro(struct xrow_header *row, char *body, const struct synchro_request *req)
 {
 	assert(iproto_type_is_synchro_request(req->type));
 
 	char *pos = body;
+	uint32_t map_size = xrow_encode_synchro_body(&pos, req);
 
-	pos = mp_encode_map(pos,
-			    iproto_type_is_promote_request(req->type) ? 3 : 2);
+	mp_encode_map(body, map_size);
 
-	pos = mp_encode_uint(pos, IPROTO_REPLICA_ID);
-	pos = mp_encode_uint(pos, req->replica_id);
+	memset(row, 0, sizeof(*row));
+	row->type = req->type;
+	row->body[0].iov_base = body;
+	row->body[0].iov_len = pos - body;
+	row->bodycnt = 1;
+}
 
-	pos = mp_encode_uint(pos, IPROTO_LSN);
-	pos = mp_encode_uint(pos, req->lsn);
+void
+xrow_encode_synchro_ownership(struct xrow_header *row, char *body,
+			      const struct synchro_ownership_request *req)
+{
+	assert(iproto_type_is_promote_request(req->base.type));
+	char *pos = body;
+	uint32_t map_size = xrow_encode_synchro_body(&pos, &req->base);
 
-	if (iproto_type_is_promote_request(req->type)) {
-		pos = mp_encode_uint(pos, IPROTO_TERM);
-		pos = mp_encode_uint(pos, req->term);
-	}
+	pos = mp_encode_uint(pos, IPROTO_TERM);
+	pos = mp_encode_uint(pos, req->term);
+	map_size += 1;
 
-	assert(pos - body < XROW_BODY_LEN_MAX);
+	mp_encode_map(body, map_size);
 
 	memset(row, 0, sizeof(*row));
-	row->type = req->type;
 	row->body[0].iov_base = body;
 	row->body[0].iov_len = pos - body;
 	row->bodycnt = 1;
 }
 
+int
+xrow_decode_synchro_body(const char *data, struct synchro_request *req)
+{
+	if (mp_typeof(*data) != MP_MAP) {
+		return -1;
+	}
+
+	memset(req, 0, sizeof(*req));
+	uint32_t map_size = mp_decode_map(&data);
+	for (uint32_t i = 0; i < map_size; i++) {
+		enum mp_type type = mp_typeof(*data);
+		if (type != MP_UINT) {
+			mp_next(&data);
+			mp_next(&data);
+			continue;
+		}
+		uint8_t key = mp_decode_uint(&data);
+		if (key < iproto_key_MAX &&
+		    iproto_key_type[key] != mp_typeof(*data)) {
+			return -1;
+		}
+		switch (key) {
+		case IPROTO_REPLICA_ID:
+			req->replica_id = mp_decode_uint(&data);
+			break;
+		case IPROTO_LSN:
+			req->lsn = mp_decode_uint(&data);
+			break;
+		default:
+			mp_next(&data);
+		}
+	}
+	return 0;
+}
+
 int
 xrow_decode_synchro(const struct xrow_header *row, struct synchro_request *req)
 {
@@ -1275,49 +1332,63 @@ xrow_decode_synchro(const struct xrow_header *row, struct synchro_request *req)
 		diag_set(ClientError, ER_INVALID_MSGPACK, "request body");
 		return -1;
 	}
-
 	assert(row->bodycnt == 1);
 
-	const char *d = (const char *)row->body[0].iov_base;
-	if (mp_typeof(*d) != MP_MAP) {
+	memset(req, 0, sizeof(*req));
+
+	if (xrow_decode_synchro_body(row->body[0].iov_base, req) != 0) {
 		xrow_on_decode_err(row, ER_INVALID_MSGPACK, "request body");
 		return -1;
 	}
 
+	req->type = row->type;
+	req->origin_id = row->replica_id;
+	return 0;
+}
+
+int
+xrow_decode_synchro_ownership(const struct xrow_header *row,
+			      struct synchro_ownership_request *req)
+{
+	if (row->bodycnt == 0) {
+		diag_set(ClientError, ER_INVALID_MSGPACK, "request body");
+		return -1;
+	}
+	assert(row->bodycnt == 1);
+
 	memset(req, 0, sizeof(*req));
-	uint32_t map_size = mp_decode_map(&d);
+
+	const char *data = row->body[0].iov_base;
+	if (xrow_decode_synchro_body(data, &req->base) != 0) {
+		xrow_on_decode_err(row, ER_INVALID_MSGPACK, "request body");
+		return -1;
+	}
+	uint32_t map_size = mp_decode_map(&data);
 	for (uint32_t i = 0; i < map_size; i++) {
-		enum mp_type type = mp_typeof(*d);
+		enum mp_type type = mp_typeof(*data);
 		if (type != MP_UINT) {
-			mp_next(&d);
-			mp_next(&d);
+			mp_next(&data);
+			mp_next(&data);
 			continue;
 		}
-		uint8_t key = mp_decode_uint(&d);
+		uint8_t key = mp_decode_uint(&data);
 		if (key < iproto_key_MAX &&
-		    iproto_key_type[key] != mp_typeof(*d)) {
+		    iproto_key_type[key] != mp_typeof(*data)) {
 			xrow_on_decode_err(row, ER_INVALID_MSGPACK,
 					   "request body");
 			return -1;
 		}
 		switch (key) {
-		case IPROTO_REPLICA_ID:
-			req->replica_id = mp_decode_uint(&d);
-			break;
-		case IPROTO_LSN:
-			req->lsn = mp_decode_uint(&d);
-			break;
 		case IPROTO_TERM:
-			req->term = mp_decode_uint(&d);
+			req->term = mp_decode_uint(&data);
 			break;
 		default:
-			mp_next(&d);
+			mp_next(&data);
 		}
 	}
 
-	req->type = row->type;
-	req->origin_id = row->replica_id;
-
+	req->base.type = row->type;
+	req->base.origin_id = row->replica_id;
 	return 0;
 }
 
diff --git a/src/box/xrow.h b/src/box/xrow.h
index fd5c4d0fc..88c3cf572 100644
--- a/src/box/xrow.h
+++ b/src/box/xrow.h
@@ -352,6 +352,12 @@ struct synchro_request {
 	 * In case of PROMOTE it means CONFIRM(lsn) + ROLLBACK(lsn+1)
 	 */
 	int64_t lsn;
+};
+
+/** Synchronous queue ownership transfer request. PROMOTE or DEMOTE. */
+struct synchro_ownership_request {
+	/** Base synchro request. */
+	struct synchro_request base;
 	/**
 	 * The new term the instance issuing this request is in. Only used for
 	 * PROMOTE request.
```

</details>

Another complication is that I have to teach `synchro_request_write` to look at the request type and interpret it either as a plain synchro_request or as a ownership_request and coll the corresponding encoder.

To sum up, IMO this doesn't simplify anything. Although I think I can make the current patch version slightly better. Applied the following diff:
```diff
diff --git a/src/box/xrow.c b/src/box/xrow.c
index 0d67c2555..89e1b62df 100644
--- a/src/box/xrow.c
+++ b/src/box/xrow.c
@@ -1245,32 +1245,32 @@ xrow_encode_synchro(struct xrow_header *row, char *body,
 
        char *pos = body;
 
-       int map_size = 2;
-       if (iproto_type_is_promote_request(req->type)) {
-               /* Term. */
-               map_size += 1;
-               /* Vclock. */
-               if (req->confirmed_vclock != NULL)
-                       map_size += 1;
-       }
-       pos = mp_encode_map(pos, map_size);
+       /* Skip one byte for the map. */
+       pos++;
+       uint32_t map_size = 0;
 
        pos = mp_encode_uint(pos, IPROTO_REPLICA_ID);
        pos = mp_encode_uint(pos, req->replica_id);
+       map_size++;
 
        pos = mp_encode_uint(pos, IPROTO_LSN);
        pos = mp_encode_uint(pos, req->lsn);
+       map_size++;
 
-       if (iproto_type_is_promote_request(req->type)) {
+       if (req->term != 0) {
                pos = mp_encode_uint(pos, IPROTO_TERM);
                pos = mp_encode_uint(pos, req->term);
-               if (req->confirmed_vclock != NULL) {
-                       pos = mp_encode_uint(pos, IPROTO_VCLOCK);
-                       pos = mp_encode_vclock_ignore0(pos,
-                                                      req->confirmed_vclock);
-               }
+               map_size++;
        }
 
+       if (req->confirmed_vclock != NULL) {
+               pos = mp_encode_uint(pos, IPROTO_VCLOCK);
+               pos = mp_encode_vclock_ignore0(pos, req->confirmed_vclock);
+               map_size++;
+       }
+
+       mp_encode_map(body, map_size);
+
        assert(pos - body < XROW_BODY_LEN_MAX);
 
        memset(row, 0, sizeof(*row));
```

Reply:
I see, thanks for the try. My idea was a bit different though - I wanted us to intentionally duplicate some code. We would have two synchro request types, yes, and they would mostly match. But this is fine as long as their encode and decode could be done via single place, like `struct replication_request` or some other base request, like you said above.

It means that this

> First of all, I'd like to keep the existing synchro request processing code intact. This means, the ownership request should be inherited from the base synchro request.

wasn't needed. Yes, there would be some duplication, but there also would be clear isolation. It would be easier to read when you see a request and you know that it corresponds to one specific type of requests. `CONFIRM` and `PROMOTE` now are diverging, so reading this new request became a little bit harder.

Anyway, I am fine with the current solution for now as well. But I will raise this topic again if we would need to diverge those request types even further from each other.

Reply:
> Yes, there would be some duplication, but there also would be clear isolation. It would be easier to read when you see a request and you know that it corresponds to one specific type of requests. CONFIRM and PROMOTE now are diverging, so reading this new request became a little bit harder.

Thanks for the explanation, I got your point better now.

Reply:
Resolving since you've approved the PR.